### ISSUE TITLE
Using a Python function for the serverLogCallback causes segfault

### ISSUE THREAD

**Issue by rudeb0t** [pymodsecurity/pymodsecurity#3]
I have been trying to use a Python function as a callback but every time Python exits, it causes a segmentation fault. I have isolated this with a test case below:

```
def cb(data, rule_msg):
    print(data, rule_msg)
    assert '[id "161"]' in rule_msg
    assert '[msg "test"]' in rule_msg

def test_log_callback1a(modsec, callback_test_rules, transaction, mocker):
    modsec.setServerLogCb(cb)

    transaction.processConnection('127.0.0.1', 33333, '127.0.0.1', 8080)
```

Running this test in isolation or in conjunction with other tests will always cause the Python interpreter to crash and dump core. Here is the backtrace from gdb session running the above test:

```
Program received signal SIGSEGV, Segmentation fault.
0x000000000810e26d in tupledealloc (op=0x7ffffa0e47d8) at Objects/tupleobject.c:239
239     Objects/tupleobject.c: No such file or directory.
(gdb) bt
#0  0x000000000810e26d in tupledealloc (op=0x7ffffa0e47d8) at Objects/tupleobject.c:239
#1  0x00000000080a9c3f in code_dealloc (co=0x7ffff1509a50) at Objects/codeobject.c:441
#2  0x00000000080c305b in func_dealloc (op=0x7ffff151bc80) at Objects/funcobject.c:527
#3  0x00007ffff9e3664b in pybind11::handle::dec_ref() const & (this=0x8c74640) at /home/nimrod/.pyenv/versions/pymodsecuritydev/include/site/python3.6/pybind11/pytypes.h:165
#4  pybind11::object::~object (this=0x8c74640, __in_chrg=<optimized out>) at /home/nimrod/.pyenv/versions/pymodsecuritydev/include/site/python3.6/pybind11/pytypes.h:208
#5  pybind11::function::~function (this=0x8c74640, __in_chrg=<optimized out>) at /home/nimrod/.pyenv/versions/pymodsecuritydev/include/site/python3.6/pybind11/pytypes.h:1212
#6  pybind11::detail::type_caster<std::function<void (char*, char const*)>, void>::load(pybind11::handle, bool)::{lambda(char*, char const*)#1}::~handle() (this=0x8c74640, __in_chrg=<optimized out>)
    at /home/nimrod/.pyenv/versions/pymodsecuritydev/include/site/python3.6/pybind11/functional.h:57
#7  std::_Function_base::_Base_manager<pybind11::detail::type_caster<std::function<void (char*, char const*)>, void>::load(pybind11::handle, bool)::{lambda(char*, char const*)#1}>::_M_destroy(std::_Any_data&, st
d::integral_constant<bool, false>) (__victim=...) at /usr/include/c++/7/bits/std_function.h:207
#8  std::_Function_base::_Base_manager<pybind11::detail::type_caster<std::function<void (char*, char const*)>, void>::load(pybind11::handle, bool)::{lambda(char*, char const*)#1}>::_M_manager(std::_Any_data&, st
d::_Function_base::_Base_manager<pybind11::detail::type_caster<std::function<void (char*, char const*)>, void>::load(pybind11::handle, bool)::{lambda(char*, char const*)#1}> const&, std::_Manager_operation) (
    __dest=..., __source=..., __op=<optimized out>) at /usr/include/c++/7/bits/std_function.h:231
#9  0x00007ffffe663041 in __run_exit_handlers (status=0, listp=0x7ffffea0b718 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#10 0x00007ffffe66313a in __GI_exit (status=<optimized out>) at exit.c:139
#11 0x000000000805ba33 in Py_Exit (sts=<optimized out>, sts@entry=0) at Python/pylifecycle.c:1546
#12 0x000000000805efbd in handle_system_exit () at Python/pythonrun.c:626
#13 0x000000000805f345 in handle_system_exit () at Python/pythonrun.c:705
#14 PyErr_PrintEx (set_sys_last_vars=1) at Python/pythonrun.c:636
#15 0x000000000805ff4c in PyErr_Print () at Python/pythonrun.c:532
#16 PyRun_SimpleFileExFlags (fp=<optimized out>, filename=<optimized out>, closeit=<optimized out>, flags=0x7ffffffede2c) at Python/pythonrun.c:425
#17 0x0000000008076e11 in run_file (p_cf=0x7ffffffede2c, filename=0x874b670 L"/home/nimrod/.pyenv/versions/3.6.5/envs/pymodsecuritydev/bin/pytest", fp=0x87993b0) at Modules/main.c:340
#18 Py_Main (argc=<optimized out>, argv=<optimized out>) at Modules/main.c:810
#19 0x000000000805627a in main (argc=4, argv=<optimized out>) at ./Programs/python.c:69
```

It seems there is a double-free from the reference count decrement on cleanup during exit. Using a `lambda` also causes a segmentation fault. For example:

```
modsec.setServerLogCb(lambda a, b: print(a, b))
```

**Comment by rudeb0t** [comment]
I wrote a quick test to see if I can replicate this issue by isolating the code that sets callback function. I cannot seem to replicate it.

Gist: https://gist.github.com/rudeb0t/5b97c652a90511ee4852b9dfbe856829

**Comment by GustavoKatel** [comment]
Thanks @rudeb0t I will try to have a look later today.

**Comment by GustavoKatel** [comment]
Hi @rudeb0t Sorry for the long delay! We just released v0.0.5 with fixes. I will close this for now and feel free to reopen if this is still happening.
